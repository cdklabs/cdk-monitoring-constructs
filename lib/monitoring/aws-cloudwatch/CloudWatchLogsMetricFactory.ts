import { DimensionsMap } from "aws-cdk-lib/aws-cloudwatch";

import {
  BaseMetricFactory,
  BaseMetricFactoryProps,
  MetricFactory,
  MetricStatistic,
} from "../../common";

const CloudWatchLogsNamespace = "AWS/Logs";

export interface CloudWatchLogsMetricFactoryProps
  extends BaseMetricFactoryProps {
  /**
   * Name of the log group to monitor.
   */
  readonly logGroupName: string;
}

export class CloudWatchLogsMetricFactory extends BaseMetricFactory<CloudWatchLogsMetricFactoryProps> {
  private readonly dimensionsMap: DimensionsMap;

  constructor(
    metricFactory: MetricFactory,
    props: CloudWatchLogsMetricFactoryProps,
  ) {
    super(metricFactory, props);

    this.dimensionsMap = {
      LogGroupName: props.logGroupName,
    };
  }

  metricIncomingLogEvents() {
    return this.metricFactory.createMetric(
      "IncomingLogEvents",
      MetricStatistic.N,
      "Logs",
      this.dimensionsMap,
      undefined,
      CloudWatchLogsNamespace,
      undefined,
      this.region,
      this.account,
    );
  }
}
