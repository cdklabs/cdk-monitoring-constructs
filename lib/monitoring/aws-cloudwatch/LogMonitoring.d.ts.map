{"version":3,"file":"LogMonitoring.d.ts","sourceRoot":"","sources":["LogMonitoring.ts"],"names":[],"mappings":"AAAA,OAAO,EAAE,OAAO,EAAkB,MAAM,4BAA4B,CAAC;AAErE,OAAO,EACL,mBAAmB,EAGnB,UAAU,EACV,eAAe,EAChB,MAAM,cAAc,CAAC;AAKtB,MAAM,WAAW,kBAAmB,SAAQ,mBAAmB;IAC7D;;OAEG;IACH,QAAQ,CAAC,YAAY,EAAE,MAAM,CAAC;IAC9B;;OAEG;IACH,QAAQ,CAAC,OAAO,EAAE,MAAM,CAAC;IACzB;;;OAGG;IACH,QAAQ,CAAC,KAAK,CAAC,EAAE,MAAM,CAAC;IACxB;;;OAGG;IACH,QAAQ,CAAC,KAAK,CAAC,EAAE,MAAM,CAAC;CACzB;AAED;;GAEG;AACH,qBAAa,aAAc,SAAQ,UAAU;IAC3C,SAAS,CAAC,QAAQ,CAAC,OAAO,EAAE,MAAM,CAAC;IACnC,SAAS,CAAC,QAAQ,CAAC,YAAY,EAAE,MAAM,CAAC;IACxC,SAAS,CAAC,QAAQ,CAAC,WAAW,CAAC,EAAE,MAAM,CAAC;IACxC,SAAS,CAAC,QAAQ,CAAC,KAAK,CAAC,EAAE,MAAM,CAAC;IAClC,SAAS,CAAC,QAAQ,CAAC,KAAK,EAAE,MAAM,CAAC;gBAErB,KAAK,EAAE,eAAe,EAAE,KAAK,EAAE,kBAAkB;IAY7D,OAAO,IAAI,OAAO,EAAE;IA0BpB,SAAS,CAAC,wBAAwB,CAAC,OAAO,EAAE,MAAM;CAKnD"}