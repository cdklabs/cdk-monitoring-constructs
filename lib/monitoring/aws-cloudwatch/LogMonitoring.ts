import {
  GraphWidget,
  HorizontalAnnotation,
  IWidget,
  LogQueryWidget,
} from "aws-cdk-lib/aws-cloudwatch";

import {
  CloudWatchLogsMetricFactory,
  CloudWatchLogsMetricFactoryProps,
} from "./CloudWatchLogsMetricFactory";
import {
  AlarmFactory,
  BaseMonitoringProps,
  CountAxisFromZero,
  DefaultGraphWidgetHeight,
  DefaultLogWidgetHeight,
  DefaultSummaryWidgetHeight,
  FullWidth,
  MaxUsageCountThreshold,
  MetricWithAlarmSupport,
  MinUsageCountThreshold,
  Monitoring,
  MonitoringScope,
  QuarterWidth,
  ThreeQuartersWidth,
  UsageAlarmFactory,
} from "../../common";
import {
  MonitoringHeaderWidget,
  MonitoringNamingStrategy,
} from "../../dashboard";

const DefaultLimit = 10;

export interface LogMonitoringProps
  extends BaseMonitoringProps,
    CloudWatchLogsMetricFactoryProps {
  /**
   * Widget title
   *
   * @default - auto-generated title based on the pattern and limit
   */
  readonly title?: string;

  /**
   * Pattern to search for, e.g. "ERROR"
   */
  readonly pattern?: string;

  /**
   * Maximum number of log messages to search for.
   *
   * @default - 10
   */
  readonly limit?: number;

  readonly addMinIncomingLogsAlarm?: Record<string, MinUsageCountThreshold>;
  readonly addMaxIncomingLogsAlarm?: Record<string, MaxUsageCountThreshold>;
}

/**
 * Monitors a CloudWatch log group for various patterns.
 */
export class LogMonitoring extends Monitoring {
  readonly logGroupName: string;
  readonly logGroupUrl?: string;

  readonly title?: string;

  readonly pattern?: string;
  readonly limit: number;

  readonly alarmFactory: AlarmFactory;
  readonly usageAlarmFactory: UsageAlarmFactory;
  readonly incomingLogEventsMetric: MetricWithAlarmSupport;

  readonly usageAnnotations: HorizontalAnnotation[];

  constructor(scope: MonitoringScope, props: LogMonitoringProps) {
    super(scope);

    this.logGroupName = props.logGroupName;
    this.logGroupUrl = scope
      .createAwsConsoleUrlFactory()
      .getCloudWatchLogGroupUrl(props.logGroupName);

    this.title = props.title;

    this.pattern = props.pattern;
    this.limit = props.limit ?? DefaultLimit;

    const namingStrategy = new MonitoringNamingStrategy({
      ...props,
      fallbackConstructName: this.logGroupName,
    });
    this.alarmFactory = this.createAlarmFactory(
      namingStrategy.resolveAlarmFriendlyName()
    );
    this.usageAlarmFactory = new UsageAlarmFactory(this.alarmFactory);

    this.usageAnnotations = [];

    const metricFactory = new CloudWatchLogsMetricFactory(
      scope.createMetricFactory(),
      props
    );
    this.incomingLogEventsMetric = metricFactory.metricIncomingLogEvents();

    for (const disambiguator in props.addMinIncomingLogsAlarm) {
      const alarmProps = props.addMinIncomingLogsAlarm[disambiguator];
      const createdAlarm = this.usageAlarmFactory.addMinUsageCountAlarm(
        this.incomingLogEventsMetric,
        alarmProps,
        disambiguator
      );
      this.usageAnnotations.push(createdAlarm.annotation);
      this.addAlarm(createdAlarm);
    }

    for (const disambiguator in props.addMaxIncomingLogsAlarm) {
      const alarmProps = props.addMaxIncomingLogsAlarm[disambiguator];
      const createdAlarm = this.usageAlarmFactory.addMaxCountAlarm(
        this.incomingLogEventsMetric,
        alarmProps,
        disambiguator
      );
      this.usageAnnotations.push(createdAlarm.annotation);
      this.addAlarm(createdAlarm);
    }

    props.useCreatedAlarms?.consume(this.createdAlarms());
  }

  summaryWidgets(): IWidget[] {
    return [
      this.createTitleWidget(),
      this.createIncomingLogsWidget(FullWidth, DefaultSummaryWidgetHeight),
    ];
  }

  widgets(): IWidget[] {
    if (this.pattern) {
      const height = this.resolveRecommendedHeight(this.limit);

      return [
        this.createTitleWidget(),

        // Log Query Results
        new LogQueryWidget({
          logGroupNames: [this.logGroupName],
          height,
          width: ThreeQuartersWidth,
          title: this.title ?? `Find ${this.pattern} (limit = ${this.limit})`,
          /**
           * https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/CWL_QuerySyntax.html
           */
          queryLines: [
            "fields @timestamp, @logStream, @message",
            `filter @message like /${this.pattern}/`,
            "sort @timestamp desc",
            `limit ${this.limit}`,
          ],
        }),

        this.createIncomingLogsWidget(QuarterWidth, height),
      ];
    } else {
      return [
        this.createTitleWidget(),
        this.createIncomingLogsWidget(FullWidth, DefaultGraphWidgetHeight),
      ];
    }
  }

  createTitleWidget() {
    return new MonitoringHeaderWidget({
      family: "Log Group",
      title: this.logGroupName,
      goToLinkUrl: this.logGroupUrl,
    });
  }

  createIncomingLogsWidget(width: number, height: number) {
    return new GraphWidget({
      width,
      height,
      title: "Incoming logs",
      left: [this.incomingLogEventsMetric],
      leftYAxis: CountAxisFromZero,
      leftAnnotations: this.usageAnnotations,
    });
  }

  protected resolveRecommendedHeight(numRows: number) {
    const heightPerLine = 1;
    const recommendedHeight = heightPerLine * numRows;
    return Math.max(recommendedHeight, DefaultLogWidgetHeight);
  }
}
