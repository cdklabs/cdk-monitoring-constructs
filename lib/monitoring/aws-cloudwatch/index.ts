export * from "./CloudWatchLogsMetricFactory";
export * from "./LogMonitoring";
