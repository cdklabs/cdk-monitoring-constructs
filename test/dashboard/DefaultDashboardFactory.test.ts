import { SynthUtils } from "@monocdk-experiment/assert";
import { Stack } from "monocdk";
import { TextWidget } from "monocdk/aws-cloudwatch";

import {
  DashboardRenderingPreference,
  DefaultDashboardFactory,
  SingleWidgetDashboardSegment,
} from "../../lib";

test("default dashboards created", () => {
  const stack = new Stack();

  new DefaultDashboardFactory(stack, "Dashboards", {
    dashboardNamePrefix: "DummyDashboard",
    createDashboard: true,
    createSummaryDashboard: true,
    createAlarmDashboard: true,
    renderingPreference: DashboardRenderingPreference.INTERACTIVE_ONLY,
  });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test("no dashboards created", () => {
  const stack = new Stack();

  const dashboards = new DefaultDashboardFactory(stack, "Dashboards", {
    dashboardNamePrefix: "DummyDashboard",
    createDashboard: false,
    createSummaryDashboard: false,
    createAlarmDashboard: false,
    renderingPreference: DashboardRenderingPreference.INTERACTIVE_ONLY,
  });

  dashboards.addSegment({
    segment: new SingleWidgetDashboardSegment(
      new TextWidget({ markdown: "Hello world!" })
    ),
  });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
