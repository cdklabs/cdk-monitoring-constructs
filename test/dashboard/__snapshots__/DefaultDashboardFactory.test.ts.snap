// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default dashboards created 1`] = `
Object {
  "Resources": Object {
    "DashboardsAlarmDashboard02C1603A": Object {
      "Properties": Object {
        "DashboardBody": "{\\"start\\":\\"-PT8H\\",\\"widgets\\":[]}",
        "DashboardName": "DummyDashboard-Alarms",
      },
      "Type": "AWS::CloudWatch::Dashboard",
    },
    "DashboardsDashboard0712AF28": Object {
      "Properties": Object {
        "DashboardBody": "{\\"start\\":\\"-PT8H\\",\\"widgets\\":[]}",
        "DashboardName": "DummyDashboard",
      },
      "Type": "AWS::CloudWatch::Dashboard",
    },
    "DashboardsSummaryDashboard1CD44475": Object {
      "Properties": Object {
        "DashboardBody": "{\\"start\\":\\"-P14D\\",\\"widgets\\":[]}",
        "DashboardName": "DummyDashboard-Summary",
      },
      "Type": "AWS::CloudWatch::Dashboard",
    },
  },
}
`;

exports[`no dashboards created 1`] = `Object {}`;
